### PR TITLE
Profile notification

### PR DESCRIPTION
**Pull request by emilianoascona**

### CODE REVIEW

File: als-server/shared/src/main/scala/org/mulesoft/als/server/modules/workspace/WorkspaceContentManager.scala [hunk 1]
```diff
@@ -247,10 +254,13 @@ class WorkspaceContentManager private (val folderUri: String,
   private def processChangeConfig(config: ProjectConfigurationState,
                                   snapshot: Snapshot,
                                   mainFileUri: Option[String]): Future[Unit] = {
+    // aca vamos a llamar al new project configuration listener: Necesitamos un validation profile manager
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/modules/workspace/WorkspaceContentManager.scala:257]: translate or remove comments in spanish

File: als-server/shared/src/main/scala/org/mulesoft/als/server/modules/ast/AstListener.scala [hunk 2]
```diff
@@ -25,6 +25,8 @@ trait AstListener[T] {
   def onRemoveFile(uri: String): Unit
 }
 
+trait WorkspaceContentListener // this just works as a "marker" for both BaseUnitListener and NewConfigurationListener
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/modules/ast/AstListener.scala:28]: why not extend `AstListener[T]`?

File: als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala [hunk 3]
```diff
@@ -35,13 +36,21 @@ class WorkspaceManager protected (val environmentProvider: EnvironmentProvider,
 
   implicit val platform: Platform = environmentProvider.platform
 
-  override def subscribers: List[BaseUnitListener] = allSubscribers.filter(_.isActive)
+  override def subscribers: List[BaseUnitListener] =
+    allSubscribers.collect({ case buL: BaseUnitListener if buL.isActive => buL })
+
+  private def workspaceSubscribers: List[AstListener[_] with WorkspaceContentListener] =
+    allSubscribers.collect({
+      case ncL: NewConfigurationListener if ncL.isActive => ncL
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala:44]: if `isActive` is part of the WorkspaceContentListener trait (or AstListener), this collect can avoid the hard match and instead just be a `filter`

File: als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala [hunk 4]
```diff
@@ -148,11 +157,14 @@ class WorkspaceList(environmentProvider: EnvironmentProvider,
                     projectConfigurationProvider: ProjectConfigurationProvider,
                     editorConfiguration: EditorConfiguration,
                     telemetryProvider: TelemetryProvider,
-                    val allSubscribers: List[BaseUnitListener],
+                    val allSubscribers: List[AstListener[_] with WorkspaceContentListener],
                     logger: Logger,
                     configurationProvider: ConfigurationProvider) {
 
-  def subscribers: List[BaseUnitListener] = allSubscribers.filter(_.isActive)
+  def subscribers: List[AstListener[_] with WorkspaceContentListener] = allSubscribers.filter(_.isActive)
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala:164]: see comment above, his filter should be a better implementation than the strict match by type

File: als-server/shared/src/test/scala/org/mulesoft/als/server/modules/project/ProfileConfigurationChangeListenerTest.scala [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  def getDiagnostics(implicit diagnosticNotifier: MockDiagnosticClientNotifier): Future[PublishDiagnosticsParams] = {
+    for {
+      _           <- diagnosticNotifier.nextCall // resolution diagnostics
+      diagnostics <- diagnosticNotifier.nextCall // custom validation diagnostics
+    } yield diagnostics
+  }
+
+  test("Should notify when a profile is added") {
+    implicit val diagnosticNotifier: MockDiagnosticClientNotifier = new MockDiagnosticClientNotifier(3000)
```
Comment (llibarona) [inline als-server/shared/src/test/scala/org/mulesoft/als/server/modules/project/ProfileConfigurationChangeListenerTest.scala:92]: where are this `diagnosticNotifier` used?

File: als-server/shared/src/test/scala/org/mulesoft/als/server/modules/project/ProfileConfigurationChangeListenerTest.scala [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    builder.build()
+  }
+
+  def buildSerializationProps(alsClient: AlsClientNotifier[StringWriter]) =
+    new SerializationProps[StringWriter](alsClient) {
+      override def newDocBuilder(prettyPrint: Boolean): DocBuilder[StringWriter] =
+        JsonOutputBuilder(prettyPrint)
+    }
+
+  def getDiagnostics(implicit diagnosticNotifier: MockDiagnosticClientNotifier): Future[PublishDiagnosticsParams] = {
```
Comment (llibarona) [inline als-server/shared/src/test/scala/org/mulesoft/als/server/modules/project/ProfileConfigurationChangeListenerTest.scala:84]: when do you need to test diagnostic in this suite?

Comment (emilianoascona) [reply als-server/shared/src/test/scala/org/mulesoft/als/server/modules/project/ProfileConfigurationChangeListenerTest.scala:84]: Nowhere, it was left here by mistake.

File: als-server/shared/src/main/scala/org/mulesoft/als/server/modules/ast/AstListener.scala (hunk 2 above)
Comment (emilianoascona) [reply als-server/shared/src/main/scala/org/mulesoft/als/server/modules/ast/AstListener.scala:28]: The original idea was to simplify without extending AstListener but given the final implementation, totally makes sense to extend

File: als-server/shared/src/main/scala/org/mulesoft/als/server/modules/workspace/WorkspaceContentManager.scala [hunk 7]
```diff
@@ -247,10 +254,12 @@ class WorkspaceContentManager private (val folderUri: String,
...
                                   snapshot: Snapshot,
                                   mainFileUri: Option[String]): Future[Unit] = {
+    val uuid = UUID.randomUUID().toString
+    configSubscribers.foreach(_.onNewAst(config, uuid))
     (mainFileUri match {
-      case Some(mainFile) if mainFile.nonEmpty => processMFChanges(mainFile, snapshot)
+      case Some(mainFile) if mainFile.nonEmpty => processMFChanges(mainFile, snapshot, uuid)
       case _                                   => Future(repository.cleanTree())
-    }).map(_ => revalidateUnits(config.profiles.map(_.path).toSet))
+    }).map(_ => revalidateUnits(config.profiles.map(_.path).toSet)) // TODO eascona: isolated profiles validation could be a config listener?
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/modules/workspace/WorkspaceContentManager.scala:262]: should this question merged? or is it a leftover?

File: als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala [hunk 8]
```diff
@@ -35,13 +35,20 @@ class WorkspaceManager protected (val environmentProvider: EnvironmentProvider,
 
   implicit val platform: Platform = environmentProvider.platform
 
-  override def subscribers: List[BaseUnitListener] = allSubscribers.filter(_.isActive)
+  override def subscribers: List[BaseUnitListener] =
+    allSubscribers.collect({ case buL: BaseUnitListener if buL.isActive => buL })
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala:39]: filter?

File: als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala [hunk 9]
```diff
@@ -35,13 +35,20 @@ class WorkspaceManager protected (val environmentProvider: EnvironmentProvider,
...
+
+  private def workspaceSubscribers: List[WorkspaceContentListener[_]] =
+    allSubscribers.collect({
+      case wcL: WorkspaceContentListener[_] if wcL.isActive => wcL
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala:43]: filter?

File: als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala [hunk 10]
```diff
@@ -148,11 +155,14 @@ class WorkspaceList(environmentProvider: EnvironmentProvider,
...
-                    val allSubscribers: List[BaseUnitListener],
+                    val allSubscribers: List[WorkspaceContentListener[_]],
                     logger: Logger,
                     configurationProvider: ConfigurationProvider) {
 
-  def subscribers: List[BaseUnitListener] = allSubscribers.filter(_.isActive)
+  def subscribers: List[WorkspaceContentListener[_]] = allSubscribers.filter(_.isActive)
+
+  private def buListenerSubscribers: List[BaseUnitListener] =
+    subscribers.collect({ case buL: BaseUnitListener => buL })
```
Comment (llibarona) [inline als-server/shared/src/main/scala/org/mulesoft/als/server/workspace/WorkspaceManager.scala:165]: filter?

**Review by llibarona — DISMISSED** [review]

**Review by llibarona — APPROVED** [review]

**Pull request merged**